execute/report/builder.go

```diff
@@ -126,47 +135,102 @@ type execReportBuilder struct {
 	addressCodec     cciptypes.AddressCodec
 
 	// Config
-	checks                []Check
-	destChainSelector     cciptypes.ChainSelector
-	maxReportSizeBytes    uint64
-	maxGas                uint64
-	maxMessages           uint64
-	maxSingleChainReports uint64
+	checks                 []Check
+	destChainSelector      cciptypes.ChainSelector
+	maxReportSizeBytes     uint64
+	maxGas                 uint64
+	maxMessages            uint64
+	maxSingleChainReports  uint64
+	multipleReportsEnabled bool
 
 	// State
-	accumulated validationMetadata
+	// accumulated keeps track of per exec report metadata as commit reports are added
+	// to the builder. It is used to limit what gets included in a report based on the
+	// builder configuration. Metadata is added to the slice as multiple reports are
+	// generated, with the last value in the slice being the currently building report.
+	accumulated []validationMetadata
 
 	// Result
-	execReports   []cciptypes.ExecutePluginReportSingleChain
-	commitReports []exectypes.CommitData
+	// execReports is the final output of the builder.
+	execReports []cciptypes.ExecutePluginReport
+	// commitReports is the updated CommitData based on the reports being built. If
+	// multiple reports are being built, it's possible for the same CommitData object
+	// to be included multiple times.
+	commitReports [][]exectypes.CommitData
+}
+
+// checkInitialize initializes builder state if needed.
+func (b *execReportBuilder) checkInitialize() {
+	if b.accumulated == nil {
+		b.accumulated = append(b.accumulated, validationMetadata{})
+	}
+	if b.execReports == nil {
+		b.execReports = append(b.execReports, cciptypes.ExecutePluginReport{})
+	}
+	if b.commitReports == nil {
+		b.commitReports = append(b.commitReports, []exectypes.CommitData{})
+	}
 }
 
 // Add an exec report for as many messages as possible in the given commit report.
-// The commit report with updated metadata is returned, it reflects which messages
+// The commit report with updated metadata is returned. It reflects which messages
 // were selected for the exec report.
+//
+// TODO: Finish this function to generate multiple reports..
 func (b *execReportBuilder) Add(
 	ctx context.Context,
 	commitReport exectypes.CommitData,
 ) (exectypes.CommitData, error) {
-	execReport, updatedReport, err := b.buildSingleChainReport(ctx, commitReport)
+	b.checkInitialize()
+	index := len(b.execReports) - 1
+
+	// Check if max is reached.
+	if b.maxSingleChainReports != 0 && len(b.execReports[index].ChainReports) >= int(b.maxSingleChainReports) {
```

Comment:
This checks the num of single-chain reports within a `Report` right? Not total number of `Report` objects (not single chain)